File: common/src/main/java/io/github/notenoughupdates/moulconfig/xml/loaders/PanelLoader.java
package io.github.notenoughupdates.moulconfig.xml.loaders;

import io.github.notenoughupdates.moulconfig.gui.component.PanelComponent;
import io.github.notenoughupdates.moulconfig.xml.ChildCount;
import io.github.notenoughupdates.moulconfig.xml.XMLContext;
import io.github.notenoughupdates.moulconfig.xml.XMLGuiLoader;
import io.github.notenoughupdates.moulconfig.xml.XMLUniverse;
import org.w3c.dom.Element;

import javax.xml.namespace.QName;
import java.util.LinkedHashMap;
import java.util.Map;

public class PanelLoader implements XMLGuiLoader.Basic<PanelComponent> {
    @Override public PanelComponent createInstance(XMLContext<?> context, Element element) {
        return new PanelComponent(
            context.getChildFragment(element),
            context.getPropertyFromAttribute(element, new QName("insets"), Integer.class, 2),
            context.getPropertyFromAttribute(element, new QName("background"), PanelComponent.BackgroundRenderer.class, PanelComponent.DefaultBackgroundRenderer.DARK_RECT)
        );
    }
    @Override public QName getName() { return XMLUniverse.qName("Panel"); }
    @Override public ChildCount getChildCount() { return ChildCount.ONE; }
    @Override public Map<String, Boolean> getAttributeNames() {
        Map<String, Boolean> map = new LinkedHashMap<>();
        map.put("insets", false);
        map.put("background", false);
        return map;
    }
}

File: common/src/main/java/io/github/notenoughupdates/moulconfig/xml/loaders/PanelLoader.kt
package io.github.notenoughupdates.moulconfig.xml.loaders

import io.github.notenoughupdates.moulconfig.gui.component.PanelComponent
import io.github.notenoughupdates.moulconfig.xml.ChildCount
import io.github.notenoughupdates.moulconfig.xml.XMLContext
import io.github.notenoughupdates.moulconfig.xml.XMLGuiLoader
import io.github.notenoughupdates.moulconfig.xml.XMLUniverse
import org.w3c.dom.Element
import javax.xml.namespace.QName

class PanelLoader : XMLGuiLoader<PanelComponent> {
    override fun createInstance(
        context: XMLContext<*>,
        element: Element
    ): PanelComponent {
        return PanelComponent(
            context.getChildFragment(
                element
            )
        )
    }

    override fun getName(): QName {
        return XMLUniverse.qName("Panel")
    }

    override fun getChildCount(): ChildCount {
        return ChildCount.ONE
    }

    override fun getAttributeNames(): Map<String, Boolean> {
        return mapOf()
    }
}